src/com/amazon/ion/impl/_Private_IonWriterBase.java

```diff
@@ -360,96 +354,128 @@ public void writeValue(IonReader reader) throws IOException
     }
 
     /**
-     * Unoptimized copy. This must not recurse back to the public
-     * {@link #writeValue(IonReader)} method since that will cause the
-     * optimization test to happen repeatedly.
+     * Writes the provided IonReader's current value including any annotations. This function will not advance the
+     * IonReader beyond the end of the current value; users wishing to continue using the IonReader at the current
+     * depth will need to call {@link IonReader#next()} again.
+     *
+     * - If the IonReader is not positioned over a value (for example: because it is at the beginning or end of a
+     *   stream), then this function does nothing.
+     * - If the current value is a container, this function will visit all of its child values and write those too,
+     *   advancing the IonReader to the end of the container in the process.
+     * - If both this writer and the IonReader are in a struct, the writer will write the current value's field name.
+     * - If the writer is not in a struct but the reader is, the writer will ignore the current value's field name.
+     * - If the writer is in a struct but the IonReader is not, this function throws an IllegalStateException.
+     *
+     * @param reader       The IonReader that will provide a value to write.
+     * @throws IOException if the provided IonReader throws an IOException.
+     * @throws IllegalStateException if this writer is inside a struct but the IonReader is not.
      */
-    final void writeValueRecursively(IonType type, IonReader reader)
-        throws IOException
+    final void writeValueRecursively(IonType ionType, IonReader reader)
```

Comment:
Hmm.... 🤨

Reply:
Zack responded offline that while the implementation is no longer recursive the value structure still is. This makes sense, traversing a recursive structure is what this does no matter how it does it.